### PR TITLE
add load slice_vars in io.py

### LINKED ISSUE #12096: ParallelExecutor CPU 单机和多机精度验证
## ParallelExecutor CPU 单机和多机精度验证
验证模型： text_classification
验证代码： https://github.com/seiriosPlus/fluid_benchmark/blob/master/text_classification/train_high_api.py

### 单机: 
CPU_NUM = 4
OPTIMIZER = SGD
BatchSzie = 160
PARALLEL_EXECUTOR = False
Epoch: 0, Step: 0, Time: 0.691313028336, Loss: 0.758364140987, Accuracy: 0.0
Epoch: 0, Step: 1, Time: 0.669323921204, Loss: 0.755404293537, Accuracy: 0.0
Epoch: 0, Step: 2, Time: 0.492468118668, Loss: 0.75151693821, Accuracy: 0.0
Epoch: 0, Step: 3, Time: 0.634160995483, Loss: 0.74986153841, Accuracy: 0.0
Epoch: 0, Step: 4, Time: 0.559371948242, Loss: 0.746825933456, Accuracy: 0.0
Epoch: 0, Step: 5, Time: 0.65730381012, Loss: 0.744488358498, Accuracy: 0.0
Epoch: 0, Step: 6, Time: 0.57714009285, Loss: 0.740698099136, Accuracy: 0.0
Epoch: 0, Step: 7, Time: 0.565196990967, Loss: 0.737674236298, Accuracy: 0.0
Epoch: 0, Step: 8, Time: 0.670403957367, Loss: 0.735484838486, Accuracy: 0.0
Epoch: 0, Step: 9, Time: 0.60733795166, Loss: 0.732566833496, Accuracy: 0.0
Epoch: 0, Step: 10, Time: 0.618324041367, Loss: 0.72943675518, Accuracy: 0.0
Epoch: 0, Step: 11, Time: 0.654896020889, Loss: 0.726955294609, Accuracy: 0.0
Epoch: 0, Step: 12, Time: 0.606182098389, Loss: 0.724970698357, Accuracy: 0.0
Epoch: 0, Step: 13, Time: 0.535058021545, Loss: 0.720089435577, Accuracy: 0.0
Epoch: 0, Step: 14, Time: 0.530606985092, Loss: 0.717955291271, Accuracy: 0.0
Epoch: 0, Step: 15, Time: 0.55975317955, Loss: 0.716057658195, Accuracy: 0.0
Epoch: 0, Step: 16, Time: 0.57083106041, Loss: 0.713347136974, Accuracy: 0.0
Epoch: 0, Step: 17, Time: 0.583375930786, Loss: 0.710104584694, Accuracy: 0.0
Epoch: 0, Step: 18, Time: 0.625988960266, Loss: 0.70750695467, Accuracy: 0.0
Epoch: 0, Step: 19, Time: 0.561584949493, Loss: 0.705342590809, Accuracy: 0.00625000009313

- - - - - 
CPU_NUM = 4
OPTIMIZER = SGD
BatchSzie = 40 (相当于每个线程是40的BatchSize)
PARALLEL_EXECUTOR = True
Epoch: 0, Step: 0, Time: 0.363101959229, Loss: 0.758364200592, Accuracy: 0.0
Epoch: 0, Step: 1, Time: 0.276148080826, Loss: 0.755404353142, Accuracy: 0.0
Epoch: 0, Step: 2, Time: 0.26346206665, Loss: 0.751516878605, Accuracy: 0.0
Epoch: 0, Step: 3, Time: 0.298815965652, Loss: 0.74986153841, Accuracy: 0.0
Epoch: 0, Step: 4, Time: 0.269979953766, Loss: 0.746825933456, Accuracy: 0.0
Epoch: 0, Step: 5, Time: 0.294218063354, Loss: 0.744488418102, Accuracy: 0.0
Epoch: 0, Step: 6, Time: 0.295897960663, Loss: 0.740698099136, Accuracy: 0.0
Epoch: 0, Step: 7, Time: 0.291954994202, Loss: 0.737674236298, Accuracy: 0.0
Epoch: 0, Step: 8, Time: 0.388696908951, Loss: 0.735484838486, Accuracy: 0.0
Epoch: 0, Step: 9, Time: 0.277570962906, Loss: 0.732566773891, Accuracy: 0.0
Epoch: 0, Step: 10, Time: 0.279649972916, Loss: 0.729436695576, Accuracy: 0.0
Epoch: 0, Step: 11, Time: 0.276800870895, Loss: 0.7269551754, Accuracy: 0.0
Epoch: 0, Step: 12, Time: 0.282086849213, Loss: 0.724970698357, Accuracy: 0.0
Epoch: 0, Step: 13, Time: 0.264585971832, Loss: 0.720089375973, Accuracy: 0.0
Epoch: 0, Step: 14, Time: 0.255345106125, Loss: 0.717955350876, Accuracy: 0.0
Epoch: 0, Step: 15, Time: 0.262812137604, Loss: 0.716057658195, Accuracy: 0.0
Epoch: 0, Step: 16, Time: 0.294218063354, Loss: 0.713347196579, Accuracy: 0.0
Epoch: 0, Step: 17, Time: 0.286861896515, Loss: 0.710104644299, Accuracy: 0.0
Epoch: 0, Step: 18, Time: 0.310491085052, Loss: 0.707507014275, Accuracy: 0.0
Epoch: 0, Step: 19, Time: 0.30881690979, Loss: 0.705342590809, Accuracy: 0.00625000009313

>  多机：
CPU_NUM = 4 
Trainer*Pserver = 1*1
BatchSize = 40 (相当于每个线程是40的BatchSize)

数据详情：
Epoch: 0, Step: 0, Time: 0.645524024963, Loss: 0.758364200592, Accuracy: 0.0
Epoch: 0, Step: 1, Time: 0.452592849731, Loss: 0.716291725636, Accuracy: 0.0
Epoch: 0, Step: 2, Time: 0.413733005524, Loss: 0.688405871391, Accuracy: 0.75
Epoch: 0, Step: 3, Time: 0.470826148987, Loss: 0.663906574249, Accuracy: 0.75
Epoch: 0, Step: 4, Time: 0.439845085144, Loss: 0.643519043922, Accuracy: 0.75
Epoch: 0, Step: 5, Time: 0.439889907837, Loss: 0.622890949249, Accuracy: 0.75
Epoch: 0, Step: 6, Time: 0.444410085678, Loss: 0.604803562164, Accuracy: 0.75
Epoch: 0, Step: 7, Time: 0.403778076172, Loss: 0.588454723358, Accuracy: 0.75
Epoch: 0, Step: 8, Time: 0.483130931854, Loss: 0.567219376564, Accuracy: 0.75
Epoch: 0, Step: 9, Time: 0.452316999435, Loss: 0.553434789181, Accuracy: 0.75
Epoch: 0, Step: 10, Time: 0.46025800705, Loss: 0.534703552723, Accuracy: 0.75
Epoch: 0, Step: 11, Time: 0.422217130661, Loss: 0.514850497246, Accuracy: 0.75
Epoch: 0, Step: 12, Time: 0.421497106552, Loss: 0.499343156815, Accuracy: 0.75
Epoch: 0, Step: 13, Time: 0.402488946915, Loss: 0.487519323826, Accuracy: 0.75
Epoch: 0, Step: 14, Time: 0.384223937988, Loss: 0.476423442364, Accuracy: 0.75
Epoch: 0, Step: 15, Time: 0.41647195816, Loss: 0.457093030214, Accuracy: 0.75
Epoch: 0, Step: 16, Time: 0.443618059158, Loss: 0.445936471224, Accuracy: 0.75
Epoch: 0, Step: 17, Time: 0.42789888382, Loss: 0.433399736881, Accuracy: 0.75
Epoch: 0, Step: 18, Time: 0.473331212997, Loss: 0.420853078365, Accuracy: 0.75
Epoch: 0, Step: 19, Time: 0.466829061508, Loss: 0.409123629332, Accuracy: 0.75

目前1x1的精度无法对齐，原因正在查找中。

### LINKED ISSUE #12201: Checkpoint Do not Save/Load block params For Distributed Fluid

### PR DESCRIPTION
**Pull request by seiriosPlus**
Need More Test.

fix #12201 
fix #12096

### CODE REVIEW

**Comment by seiriosPlus** [comment]
Need Add unit tests.

**Comment by seiriosPlus** [comment]
related #12167

**Comment by seiriosPlus** [comment]
related #12163

File: python/paddle/fluid/io.py [hunk 1]
```diff
@@ -790,3 +794,46 @@ def get_parameter_value_by_name(name, executor, program=None):
         program = default_main_program()
     var = program.global_block().var(name)
     return get_parameter_value(var, executor)
+
+
+def _load_slice_up_vars(executor, dirname, slice_vars_and_atts):
+    if slice_vars_and_atts == None or len(slice_vars_and_atts) == 0:
```
Comment (panyx0718) [inline python/paddle/fluid/io.py]: ```if not slice_vars_and_atts```

File: python/paddle/fluid/trainer.py [hunk 2]
```diff
@@ -1043,9 +1087,9 @@ def _save_trainer_args(dirname, trainer_id, trainer_args):
     _write_success(cur_dir)
 
 
-def _load_trainer_args(checkpoint_dir, serial, trainer_id, trainer_args):
+def _load_trainer_args(checkpoint_dir, trainer_id, trainer_args):
```
Comment (panyx0718) [inline python/paddle/fluid/trainer.py]: update "Args" below?

File: python/paddle/fluid/io.py [hunk 3]
```diff
@@ -742,6 +763,118 @@ def load_inference_model(dirname,
...
+
+    LOOKUP_TABLE_TYPE = "lookup_table"
+    lookup_table_var_name = None
+
+    for op in program.global_block().ops:
+        if op.type == LOOKUP_TABLE_TYPE:
+            if op.attrs['is_distributed'] is True:
+                if lookup_table_var_name is None:
+                    lookup_table_var_name = op.input("W")[0]
+                if lookup_table_var_name != op.input("W")[0]:
```
Comment (panyx0718) [inline python/paddle/fluid/io.py]: elif?

File: python/paddle/fluid/io.py [hunk 4]
```diff
@@ -742,6 +763,118 @@ def load_inference_model(dirname,
...
+                    raise RuntimeError("all distributed lookup_table_ops"
+                                       " should have only one table")
+
+    lookup_table_var = program.global_block().vars[lookup_table_var_name]
```
Comment (panyx0718) [inline python/paddle/fluid/io.py]: Does this return None if not exist?

File: python/paddle/fluid/io.py [hunk 5]
```diff
@@ -742,6 +763,120 @@ def load_inference_model(dirname,
...
+
+    executor.run(load_prog)
+
+
+def _endpoints_replacement(program, endpoints):
+    ENDPOINT_MAP = "epmap"
+    for op in program.global_block().ops:
+        if op.has_attr(ENDPOINT_MAP):
+            op.set_attr(ENDPOINT_MAP, endpoints)
+    program = program.clone()
```
Comment (panyx0718) [inline python/paddle/fluid/io.py]: Do you want to mutate after clone or clone after mutate?

**Review by panyx0718 — COMMENTED** [review]
why are there so many changes in trainer.py? 

there seems to be no test at all?

File: python/paddle/fluid/io.py (hunk 5 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: use ```sync_cpp``` to replace it.

File: python/paddle/fluid/io.py (hunk 1 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

**Comment by seiriosPlus** [comment]
I will open a new pull request to push  ``` trianer.py```.

File: paddle/fluid/API.spec [hunk 6]
```diff
@@ -77,8 +77,8 @@ paddle.fluid.io.save_persistables ArgSpec(args=['executor', 'dirname', 'main_pro
 paddle.fluid.io.load_vars ArgSpec(args=['executor', 'dirname', 'main_program', 'vars', 'predicate', 'filename'], varargs=None, keywords=None, defaults=(None, None, None, None))
 paddle.fluid.io.load_params ArgSpec(args=['executor', 'dirname', 'main_program', 'filename'], varargs=None, keywords=None, defaults=(None, None))
 paddle.fluid.io.load_persistables ArgSpec(args=['executor', 'dirname', 'main_program', 'filename'], varargs=None, keywords=None, defaults=(None, None))
+paddle.fluid.io.load_inference_model ArgSpec(args=['dirname', 'executor', 'model_filename', 'params_filename', 'pserver_endpoints'], varargs=None, keywords=None, defaults=(None, None))
```
Comment (jacquesqiao) [inline paddle/fluid/API.spec]: 这个顺序改变不会影响测试么？

File: python/paddle/fluid/io.py [hunk 7]
```diff
@@ -667,11 +671,20 @@ def save_inference_model(dirname,
 
     save_persistables(executor, dirname, inference_program, params_filename)
 
+    # if there is lookup table, the trainer 0 will notify all pserver to save.
+    if main_program._is_distributed and main_program._is_chief:
+        if main_program._distributed_lookup_table:
```
Comment (panyx0718) [inline python/paddle/fluid/io.py]: combine the 2 if?

**Review by panyx0718 — COMMENTED** [review]
LG! Can you add a test?

File: python/paddle/fluid/io.py (hunk 7 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: paddle/fluid/operators/distributed/request_handler_impl.cc [hunk 8]
```diff
@@ -130,12 +130,12 @@ bool RequestCheckpointHandler::Handle(const std::string& varname,
       checkpoint_notify_id != -1,
       "when checkpoint_notify_id = -1, there should be no RPC invoke.");
 
-  auto* lt_var = scope->FindVar(LOOKUP_TABLE_PATH)->GetMutable<std::string>();
+  auto* lt_var = scope_->FindVar(LOOKUP_TABLE_PATH)->GetMutable<std::string>();
```
Comment (jacquesqiao) [inline paddle/fluid/operators/distributed/request_handler_impl.cc]: add TODO here to find why scope will be error.

**Review by jacquesqiao — DISMISSED** [review]
LGTM!

**Review by typhoonzero — DISMISSED** [review]
LGTM for API change!

File: paddle/fluid/operators/distributed/request_handler_impl.cc [hunk 9]
```diff
@@ -130,12 +130,13 @@ bool RequestCheckpointHandler::Handle(const std::string& varname,
       checkpoint_notify_id != -1,
       "when checkpoint_notify_id = -1, there should be no RPC invoke.");
 
-  auto* lt_var = scope->FindVar(LOOKUP_TABLE_PATH)->GetMutable<std::string>();
+  // TODO(tangwei12): find out why scope will be error.
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed/request_handler_impl.cc:133]: why?

File: python/paddle/fluid/tests/unittests/test_dist_transpiler.py [hunk 10]
```diff
@@ -582,6 +584,23 @@ def transpiler_test_impl(self):
         self.assertEqual(row_size, calc_row_size)
 
 
+class TestDistArgsInProgram(TestDistLookupTableBase):
+    def net_conf(self):
+        self.network_with_table(is_sparse=True, is_distributed=True)
+
+    def transpiler_test_impl(self):
+        config = fluid.DistributeTranspilerConfig()
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_dist_transpiler.py]: what is this for?

File: python/paddle/fluid/tests/unittests/test_dist_transpiler.py [hunk 11]
```diff
@@ -612,5 +631,41 @@ def transpiler_test_impl(self):
...
+                                    size=1000,
+                                    act=None,
+                                    param_attr=fluid.ParamAttr(name='fc_w'),
+                                    bias_attr=fluid.ParamAttr(name='fc_b'))
+        y = fluid.layers.data(name='y', shape=[1], dtype='float32')
+        cost = fluid.layers.square_error_cost(input=y_predict, label=y)
+        avg_cost = fluid.layers.mean(cost)
+        optimizer = fluid.optimizer.RMSProp(learning_rate=0.1)
+        optimizer.minimize(avg_cost)
+        return
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_dist_transpiler.py]: no need?

File: python/paddle/fluid/tests/unittests/test_dist_transpiler.py [hunk 12]
```diff
@@ -582,6 +584,23 @@ def transpiler_test_impl(self):
         self.assertEqual(row_size, calc_row_size)
 
 
+class TestDistArgsInProgram(TestDistLookupTableBase):
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_dist_transpiler.py:582]: I don't trust this kind of transpiler test.
I want to see a dist look up table that is actually saved and loaded by executor.

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 13]
```diff
@@ -689,8 +702,35 @@ def _get_splited_name_and_shape(varname):
                     inputs=new_inputs,
                     outputs=new_outputs,
                     attrs=op.all_attrs())
+
+        # add slice vars
+        slice_vars_and_atts = self._get_slice_vars_and_atts(endpoint)
+        s_prog._slice_vars_and_atts = slice_vars_and_atts
```
Comment (panyx0718) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: why make this 2 lines?

**Review by panyx0718 — DISMISSED** [review]

File: python/paddle/fluid/tests/unittests/test_dist_transpiler.py (hunk 10 above)
Comment (seiriosPlus) [reply python/paddle/fluid/tests/unittests/test_dist_transpiler.py]: done

File: python/paddle/fluid/tests/unittests/test_dist_transpiler.py (hunk 11 above)
Comment (seiriosPlus) [reply python/paddle/fluid/tests/unittests/test_dist_transpiler.py]: done

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 13 above)
Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: done

**Review by panyx0718 — DISMISSED** [review]

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 14]
```diff
@@ -689,8 +702,34 @@ def _get_splited_name_and_shape(varname):
...
+        s_prog._slice_vars_and_atts = self._get_slice_vars_and_atts(endpoint)
+
         return s_prog
 
+    def _get_slice_vars_and_atts(self, endpoint):
+        slice_vars_and_atts = []
+        block_suffix = ".block"
+        for param in self.param_grad_ep_mapping[endpoint]["params"]:
+
+            suff_idx = param.name.find(block_suffix)
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: use `_get_varname_parts` here.

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 15]
```diff
@@ -689,8 +702,34 @@ def _get_splited_name_and_shape(varname):
                     inputs=new_inputs,
                     outputs=new_outputs,
                     attrs=op.all_attrs())
+
+        # add slice vars
+        s_prog._slice_vars_and_atts = self._get_slice_vars_and_atts(endpoint)
+
         return s_prog
 
+    def _get_slice_vars_and_atts(self, endpoint):
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: `_get_slice_vars_and_attrs`

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 16]
```diff
@@ -689,8 +702,34 @@ def _get_splited_name_and_shape(varname):
                     inputs=new_inputs,
                     outputs=new_outputs,
                     attrs=op.all_attrs())
+
+        # add slice vars
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py:706]: comment not needed

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 14 above)
Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: done

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 15 above)
Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: done

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 16 above)
Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py:706]: done

**Comment by typhoonzero** [comment]
need to update API.spec

**Review by typhoonzero — APPROVED** [review]
LGTM++

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**